internal/telemetry/telemetry.go

```diff
@@ -0,0 +1,382 @@
+// Copyright 2021-2022 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"runtime"
+	"runtime/debug"
+	"strconv"
+	"time"
+
+	statev1 "github.com/cerbos/cerbos/api/genpb/cerbos/state/v1"
+	telemetryv1 "github.com/cerbos/cerbos/api/genpb/cerbos/telemetry/v1"
+	"github.com/cerbos/cerbos/internal/audit"
+	"github.com/cerbos/cerbos/internal/config"
+	"github.com/cerbos/cerbos/internal/schema"
+	"github.com/cerbos/cerbos/internal/storage"
+	"github.com/cerbos/cerbos/internal/storage/blob"
+	"github.com/cerbos/cerbos/internal/storage/disk"
+	"github.com/cerbos/cerbos/internal/storage/git"
+	"github.com/cerbos/cerbos/internal/util"
+	"github.com/google/uuid"
+	"github.com/spf13/afero"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	analytics "gopkg.in/segmentio/analytics-go.v3"
+)
+
+const (
+	doNotTrackEnvVar  = "DO_NOT_TRACK"
+	noTelemetryEnvVar = "CERBOS_NO_TELEMETRY"
+
+	minReportInterval = 8 * time.Hour
+	stateFile         = "cerbos.telemetry.json"
+)
+
+var SegmentWriteKey string
+
+func Report(ctx context.Context, store storage.Store) {
+	logger := zap.L().Named("telemetry")
+	if SegmentWriteKey == "" {
+		logger.Info("Telemetry disabled")
+		return
+	}
+
+	conf := &Conf{}
+	_ = config.GetSection(conf)
+
+	if conf.Disabled {
+		logger.Info("Telemetry disabled")
+		return
+	}
+
+	go doReport(ctx, store, conf, logger)
+}
+
+func doReport(ctx context.Context, store storage.Store, conf *Conf, logger *zap.Logger) {
+	defer func() {
+		// don't let a panic in this goroutine crash the whole app.
+		if err := recover(); err != nil {
+			logger.Debug("Telemetry panic", zap.Any("cause", err))
+		}
+	}()
+
+	fs := initStateFS(conf.StateDir)
+	r := newReporter(store, fs, logger)
+	r.report(ctx)
+}
+
+func initStateFS(dir string) afero.Fs {
+	if dir == "" {
+		confDir, err := os.UserConfigDir()
+		if err != nil {
+			confDir = os.TempDir()
+		}
+
+		dir = filepath.Join(confDir, util.AppName)
+	}
+
+	finfo, err := os.Stat(dir)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return afero.NewMemMapFs()
+		}
+
+		//nolint:gomnd
+		if err := os.MkdirAll(dir, 0o700); err != nil {
+			return afero.NewMemMapFs()
+		}
+	}
+
+	if finfo != nil && !finfo.IsDir() {
+		return afero.NewMemMapFs()
+	}
+
+	return afero.NewBasePathFs(afero.NewOsFs(), dir)
+}
+
+type reporter struct {
+	fsys   afero.Fs
+	store  storage.Store
+	logger *zap.Logger
+}
+
+func newReporter(store storage.Store, fsys afero.Fs, logger *zap.Logger) *reporter {
+	return &reporter{
+		fsys:   fsys,
+		store:  store,
+		logger: logger,
+	}
+}
+
+func (r *reporter) report(ctx context.Context) bool {
+	state := r.readState()
+	if state == nil {
```

Comment:
Looking at `readState`, I don't think it's possible for `state` to be `nil` at this point.